### ISSUE TITLE
Make pressing tab in script editor always increase indentation

### ISSUE THREAD

**Issue by jmdejong** [godotengine/godot-proposals#7816]
### Describe the project you are working on

Any project where I use the scripting editor

### Describe the problem or limitation you are having in your project

When pressing the tab key with the cursor in the middle of the line a new tab character will appear in the middle of the line, instead of indenting the code.

Indenting the code is something very common. Inserting a tab in the middle of a line is very uncommon (I guess the only use case is to put a tab inside a string, but I've never had to do that).
Many other code editors will increase the indentation when the tab key is pressed regardless of where the cursor is.
The current behaviour is also inconsistent with shift+tab: first pressing shift+tab and then tab will decrease the indentation level and then insert a tab in the middle of the line.
The current behaviour is also inconsistent with pressing the tab key while some text is selected. If the tab key is pressed while some text is selected then the indentation is always increased. For any other character the selected text would be replaced with that character

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Pressing the tab key should increase the indentation of the current line instead of inserting a tab character at the position of the cursor

If there would be a good reason to insert a tab character at the position of the cursor then there should be a toggle in the settings to switch between the current behaviour and the proposed behaviour

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

When the user presses the tab key and the code editor is selected, then the current line should be indented one more level

or

When the user presses the tab key the code editor is selected, and the setting to always indent on tab press is on, then the current line should be indented one more level

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This can be worked around by using an external editor, but that would miss the quick feedback on the code.

It is also possible to always go to the start of the line or to select some text when indenting, but that is quite inconvenient.

### Is there a reason why this should be core and not an add-on in the asset library?

This is a common feature in many code editors, and I think that the godot code editor ought to have this feature too to be a good code editor

**Comment by Zireael07** [comment]
Counter argument: I often have tab characters in my strings

**Comment by jmdejong** [comment]
Interesting, I did not expect that to be common. What do you use them for?

**Comment by Zireael07** [comment]
Usually for visually aligning stuff, but sometimes I have actual paragraphs of text (in a RichTextLabel) because it's a scroll or a book in-game

**Comment by AThousandShips** [comment]
> When pressing the tab key with the cursor in the middle of the line a new tab character will appear in the middle of the line, instead of indenting the code.

This happens in many editors, like Kate for example, I don't think this is unusual but rather the norm

**Comment by jmdejong** [comment]
> Usually for visually aligning stuff, but sometimes I have actual paragraphs of text (in a RichTextLabel) because it's a scroll or a book in-game

This would not apply to the editor for a RichTextLabel because that is not a code editor. Code editors already have different behaviour for how tabs are handled: pressing tab while text is selected in a code editor will increase indentation, while pressing tab while text is selected in another type of text editor will replace the selected text with a tab character

**Comment by jmdejong** [comment]
> > When pressing the tab key with the cursor in the middle of the line a new tab character will appear in the middle of the line, instead of indenting the code.
> 
> This happens in many editors, like Kate for example, I don't think this is unusual but rather the norm

I just checked and it's a setting in kate:

![kate_indentation_settings](https://github.com/godotengine/godot-proposals/assets/7968950/534ef51c-e12b-46c4-a4b0-7ea8460150ba)

I always have it to "always increase indentation level" and I assumed that was the default, but I'm not actually sure if that's true.

**Comment by Zireael07** [comment]
@jmdejong Yes I know but I often have the strings in script and set RTL content from said script

That said, a setting like you show for Kate would probably be the best solution.

**Comment by jmdejong** [comment]
The current behaviour of the code editor in godot seems to correspond with "Increase indentation level if in leading blank space".
I think we could leave out the "Always advance to the next tab position" behaviour. As far as I could see it will differ only with the current behaviour would be if you use spaces for indentation by allowing you to easily keep an indentation that is not a multiple of the indetation size (not a style I would recommend).
This would mean that for godot there would just be 2 options to choose from.
Are there any counter points to this?
